libp2p-xtra/src/node.rs

```diff
@@ -0,0 +1,508 @@
+use crate::multiaddress_ext::MultiaddrExt as _;
+use crate::upgrade;
+use crate::Connection;
+use crate::Substream;
+use anyhow::bail;
+use anyhow::Context as _;
+use anyhow::Result;
+use async_trait::async_trait;
+use futures::future::BoxFuture;
+use futures::stream::BoxStream;
+use futures::AsyncRead;
+use futures::AsyncWrite;
+use futures::TryStreamExt;
+use libp2p_core::identity::Keypair;
+use libp2p_core::transport::Boxed;
+use libp2p_core::transport::ListenerEvent;
+use libp2p_core::Multiaddr;
+use libp2p_core::PeerId;
+use libp2p_core::Transport;
+use multistream_select::NegotiationError;
+use multistream_select::Version;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::marker::PhantomData;
+use std::time::Duration;
+use thiserror::Error;
+use tokio_tasks::Tasks;
+use xtra::message_channel::StrongMessageChannel;
+use xtra_productivity::xtra_productivity;
+use xtras::SendAsyncSafe;
+
+/// An actor for managing multiplexed connections over a given transport.
+///
+/// The actor does not impose any policy on connection and/or protocol management.
+/// New connections can be established by sending a [`Connect`] messages. Existing connections can
+/// be disconnected by sending [`Disconnect`]. Listening for incoming connections is done by sending
+/// a [`ListenOn`] message. To list the current state, send the [`GetConnectionStats`] message.
+///
+/// The combination of the above should make it possible to implement a fairly large number of
+/// policies. For example, to maintain a connection to a specific node, you can regularly check if
+/// the connection is still established by sending [`GetConnectionStats`] and react accordingly
+/// (f.e. sending [`Connect`] in case the connection has disappeared).
+///
+/// Once a connection with a peer is established, both sides can open substreams on top of the
+/// connection. Any incoming substream will - assuming the protocol is supported by the node -
+/// trigger a [`NewInboundSubstream`] message to the actor provided in the constructor.
+/// Opening a new substream can be achieved by sending the [`OpenSubstream`] message.
+pub struct Node {
```

Comment:
would you be open to renaming this super-generic actor to something better? `ConnectionManager`, `Connector`, something along these lines? Open to other ones too, the current name makes it hard to reason IMHO

Reply:
Fair criticism. Looking at some prior art, QUIC implemenations name this thing an `Endpoint`. Thoughts on that?

I think `ConnectionManager` is not a great name because it does much more than manage connections for you. IMO, the core feature of this component is the ability to obtain a multiplexed, virtual stream to the other party. The fact that a physical connection is required for that is explicitly moved into the background.

Because of this abstraction, I named it `Node`. Similar to how a Bitcoin Node gives you a relatively easy to use interface to the Bitcoin network. Node also has this co-notation of a p2p network which I wanted to express.

Some names that come to my mind (none of which I am super happy with, but maybe they help the brain-storming).

- `StreamPortal` (a bit funky)
- `Network` (too broad)
- `Endpoint` (as suggested above)
- `NetworkStreamManager` (a bit long)
- `PeerManager` (I am not a massive fan of "Manager" names haha)

Reply:
I like `Endpoint` much more! I can get convinced by the prior art argument. 
`StreamPortal` is also good, as it describes essentially what it does

Reply:
Renamed to `Endpoint`!